### PR TITLE
Chore: Report frontend metrics for JS/CSS load time

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

Grafana backend stores the following frontend metrics in Prometheus data base [(frontendmetrics.go)](https://github.com/grafana/grafana/blob/main/pkg/infra/metrics/frontendmetrics.go#L39):
* `frontend_boot_load_time_seconds`
* `frontend_boot_first_paint_time_seconds`
* `frontend_boot_js_done_time_seconds`

However metrics that are sent from the frontend are:

Screen Shot 2021-11-29 at 10 43 43

Meaning:
* `frontend_boot_js_done_time_seconds` is not passed from the frontend
* first contentful paint is passed from the frontend but not used in the backend
* css load time (measured in [index-template.htm](https://github.com/grafana/grafana/blob/main/public/views/index-template.html#L29)) is not used

This PR adds:
* Storing additional FCP metric (`frontend_boot_first_contentful_paint_time_seconds`) 
* Passing `frontend_boot_js_done_time_seconds` and `frontend_boot_css_time_seconds`

**Which issue(s) this PR fixes**:

Fixes N/A

**Special notes for your reviewer**:

_How to test it?_ Use devenv prometheus (`make run devenv sources=prometheus`) and query for metrics reported in `frontendmetrics.go`.

Screen Shot 2021-11-29 at 11 03 53

### CODE REVIEW

File: public/app/app.ts [hunk 1]
```diff
@@ -179,8 +182,9 @@ function initEchoSrv() {
         );
       }
 
-      const loadMetric = performance.getEntriesByName(loadMetricName)[0];
-      reportPerformance(loadMetric.name, Math.round(loadMetric.startTime + loadMetric.duration) / 1000);
+      reportMetricPerformance(loadMetricName);
```
Comment (aocenas) [inline public/app/app.ts:185]: What is this actually measuring? We mark it on line 174 and then report it here. It seems like it is only measuring how long it takes to report the paint metrics?

File: public/app/app.ts [hunk 2]
```diff
@@ -166,6 +166,9 @@ function initEchoSrv() {
 
   window.addEventListener('load', (e) => {
     const loadMetricName = 'frontend_boot_load_time_seconds';
+    // should match names in public/views/index-template.html
+    const jsLoadMetricName = 'frontend_boot_js_done_time_seconds';
+    const cssLoadMetricName = 'frontend_boot_css_time_seconds';
 
     if (performance && performance.getEntriesByType) {
```
Comment (aocenas) [inline public/app/app.ts:173]: This is bit confusing. To actually measure anything we have to explicitly report it but also backend needs to know about the measurement. So gathering it dynamically like this does not make sense to me. If browser starts reporting some more metrics we wouldn't automatically add it to prometheus because the histogram isn't registered. So I think it would make sense to just statically write down the metrics here for the paints and comment that it needs to match the `frontendmetrics.go` registered histograms.

**Review by torkelo — APPROVED** [review]

File: public/app/app.ts [hunk 3]
```diff
@@ -166,6 +166,9 @@ function initEchoSrv() {
 
   window.addEventListener('load', (e) => {
     const loadMetricName = 'frontend_boot_load_time_seconds';
+    // should match names in public/views/index-template.html
```
Comment (aocenas) [inline public/app/app.ts:169]: nit: I would reword this to say these are 'marked' in public/views/index-template.html to keep with the `performance` API naming.

**Review by aocenas — COMMENTED** [review]
Left some comments which may not be really about the particular changes here but thought they are important as the code confused me quite a bit.

File: public/app/app.ts (hunk 2 above)
Comment (ifrost) [reply public/app/app.ts:173]: It measures the time up until that line is executed - from initial page load to window load event.

File: public/app/app.ts (hunk 1 above)
Comment (ifrost) [reply public/app/app.ts:185]: It measures the time up until that line is executed - from initial page load to window load event.

File: public/app/app.ts (hunk 2 above)
Comment (ifrost) [reply public/app/app.ts]: Makes sense. I'll change them to be statically defined in frontend.

File: public/app/app.ts (hunk 3 above)
Comment (ifrost) [reply public/app/app.ts]: Updated.

**Comment by ifrost** [comment]
A couple of changes:

I'm not sure why we use `(metric.startTime + metric.duration)`. From docs it seems that `metric.startTime` should be the timestamp from [the time origin](https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin), which should be enough to report at what point in document lifecycle the event happened.

What is more `metric.duration` should always be `0`  in our cases because:
- `performance.mark()` entries have duration of 0 (https://developer.mozilla.org/en-US/docs/Web/API/PerformanceMark)
- `first-paint` and `first-contentful-paint` are of type PerformancePaintTiming  which also has duration always set to `0` (https://developer.mozilla.org/en-US/docs/Web/API/PerformancePaintTiming)

I have changed `(metric.startTime + metric.duration)` to just `metric.startTime`. Metrics are also defined statically in frontend. I've also moved converting `-` to `_` in metric names from backend to frontend.

**Review by aocenas — APPROVED** [review]
So this makes more sense to me, but maybe somebody who added this before could also take a look?

**Comment by ifrost** [comment]
>  So this makes more sense to me, but maybe somebody who added this before could also take a look?

@torkelo - I've made some changes based on @aocenas comments. Could you please check if they make sense?

**Review by torkelo — APPROVED** [review]
